src/umpire/util/AllocationMap.hpp

```diff
@@ -12,72 +12,106 @@
+  const AllocationRecord* operator->() const;
+  AllocationMapConstIterator& operator++();
+  AllocationMapConstIterator operator++(int);
+
+  bool operator==(const AllocationMapConstIterator& other);
+  bool operator!=(const AllocationMapConstIterator& other);
+private:
+  Judy* m_array;
+  JudySlot* m_last;
+  uintptr_t m_ptr;
+  RecordList* m_list;
+  RecordListConstIterator* m_iter;
+};
 
-    class ConstIterator {
-    public:
-      const AllocationRecord* operator*();
-      ConstIterator& operator++();
-      bool operator==(const ConstIterator& other);
-      bool operator!=(const ConstIterator& other);
-    private:
-      struct JudyL2Data;
-      bool end;
-      JudyL2Data* data;
-      ConstIterator(judyL2Array<uintptr_t, uintptr_t>* map_, const bool end_ = false);
-      friend class AllocationMap;
-    };
+class AllocationMap
+{
+public:
+  // Friend the iterator class
+  friend class RecordListConstIterator;
+  friend class AllocationMapConstIterator;
 
-    AllocationMap();
-    ~AllocationMap();
+  AllocationMap();
+  ~AllocationMap();
 
-    void insert(void* ptr, AllocationRecord* record);
+  // Would require a deep copy of the Judy data
+  AllocationMap(const AllocationMap&) = delete;
 
-    AllocationRecord* remove(void* ptr);
+  // Insert a new record -- copies record
+  void insert(void* ptr, AllocationRecord record);
 
-    AllocationRecord* find(void* ptr) const;
+  // Find a record -- throws an exception of the record is not found
+  AllocationRecord* find(void* ptr) const;
 
-    ConstIterator begin() const;
+  // This version of find never throws an exception
+  AllocationRecord* findRecord(void* ptr) const;
 
-    ConstIterator end() const;
+  // Only allows erasing the last inserted entry for key = ptr
+  AllocationRecord remove(void* ptr);
 
-    bool contains(void* ptr);
+  // Check if a pointer has been added to the map
+  bool contains(void* ptr) const;
 
-    void reset();
+  // Clear all records from the map
+  void clear();
 
-    void printAll(std::ostream& os = std::cout) const;
+  // Number of entries
+  size_t size() const;
 
-    void print(const std::function<bool (const AllocationRecord*)>&& predicate,
-               std::ostream& os = std::cout) const;
+  // Print methods -- either matching a predicate or all records
+  void print(const std::function<bool (const AllocationRecord&)>&& predicate,
+             std::ostream& os = std::cout) const;
 
-private:
-    AllocationRecord* findRecord(void* ptr) const;
+  void printAll(std::ostream& os = std::cout) const;
 
-    // TODO: Make const version of judyL2Array begin/end
-    mutable judyL2Array<uintptr_t, uintptr_t>* m_records;
+  // Const iterator
+  AllocationMapConstIterator begin() const;
+  AllocationMapConstIterator end() const;
 
-    std::mutex* m_mutex;
+private:
+  Judy* m_array;
+  mutable JudySlot* m_last; // last found value in Judy
+  unsigned int m_max_levels, m_depth;
+  size_t m_size;
+  std::mutex* m_mutex;
```

Comment:
Can we remove the indirection from mutex locks `UMPIRE_LOCK / UMPIRE_UNLOCK`?

Reply:
I'm no sure.  Why do you wish to do so?  @davidbeckingsale, @jdahm?

Reply:
Fight the indirection